src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java

```diff
@@ -52,5 +293,404 @@ public void execute() throws MojoExecutionException, MojoFailureException {
+                String key = entry.getKey();
+                String[] groupArt = key.split(":");
+                String groupId = groupArt[0];
+                String artifactId = groupArt[1];
+                String version = entry.getValue();
+                /*
+                 * We cannot use MavenProject.getArtifactMap since we may have multiple dependencies
+                 * of different classifiers.
+                 */
+                boolean found = false;
+                // Yeah, this is O(n²)... deal with it!
+                for (MavenArtifact a : effectiveArtifacts) {
+                    if (!a.getGroupId().equals(groupId) || !a.getArtifactId().equals(artifactId)) {
+                        continue;
+                    }
+                    if (!a.getVersion().equals(version)) {
+                        throw new AssertionError("should never happen");
+                    }
+                    found = true;
+                    if (a.getArtifactHandler().isAddedToClasspath()) {
+                        /*
+                         * Everything is added to the test classpath, so there is no need to check
+                         * scope.
+                         */
+                        additionalClasspathElements.add(a.getFile().getAbsolutePath());
+                    }
+                }
+                if (!found) {
+                    throw new MojoExecutionException("could not find dependency " + key);
+                }
+            }
+
+            NavigableSet<String> classpathDependencyExcludes = new TreeSet<>();
+            classpathDependencyExcludes.addAll(deletions.keySet());
+            classpathDependencyExcludes.addAll(updates.keySet());
+
+            Properties properties = project.getProperties();
+            if (getLog().isDebugEnabled()) {
+                getLog().debug(String.format("Replacing POM-defined classpath elements %s with %s", classpathDependencyExcludes, additionalClasspathElements));
+            }
+            // cf. http://maven.apache.org/surefire/maven-surefire-plugin/test-mojo.html
+            properties.setProperty("maven.test.additionalClasspath", String.join(",", additionalClasspathElements));
+            properties.setProperty("maven.test.dependency.excludes", String.join(",", classpathDependencyExcludes));
+        }
+    }
+
+    /**
+     * Scan a WAR file, accumulating plugin information.
+     *
+     * @param war The WAR to scan.
+     * @return The bundled plugins in the WAR.
+     */
+    @SuppressFBWarnings(value = "REDOS", justification = "trusted code")
+    private static Map<String, String> scanWar(File war, MavenSession session, MavenProject project) throws MojoExecutionException {
+        Map<String, String> overrides = new HashMap<>();
+        try (JarFile jf = new JarFile(war)) {
+            Enumeration<JarEntry> entries = jf.entries();
+            String coreVersion = null;
+            while (entries.hasMoreElements()) {
```

Comment:
Patterns are thread safe and compilation is not free so should not be recompiled in loops.

```suggestion
            final Pattern corePattern = Pattern.compile(CORE_REGEX);
            final Pattern pluginPattern = Pattern.compile(PLUGIN_REGEX);
            while (entries.hasMoreElements()) {
```

Reply:
This code was copied from `plugin-compat-tester`. I note that you are comfortable pointing out the flaw in the context of this PR but not comfortable opening a pull request to fix the issue in `plugin-compat-tester`.

Reply:
I did not know this came from the PCT.